Limit how many concurrent integration tests run.

bowlofeggs opened the pull request:
I found that on my laptop, the integration tests would often fail
when run with bodhi-ci's default -j flag in a Vagrant guest. I
discovered that the failures were very often due to timeouts in
operations on the postgresql database, and I also discovered that
running bodhi-ci with a small -j flag (such as -j 2) would get them
to pass. Rather than defaulting to a smaller -j flag value, this
commit adds a second concurrency semaphore for IntegrationJobs to
limit how many of them can run at once.

The chosen value is admittedly a bit "magic", but was chosen
because it happens to get the tests to pass on my Lenovo T480s with
a Vagrant guest.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs committed:
Limit how many concurrent integration tests run.

I found that on my laptop, the integration tests would often fail
when run with bodhi-ci's default -j flag in a Vagrant guest. I
discovered that the failures were very often due to timeouts in
operations on the postgresql database, and I also discovered that
running bodhi-ci with a small -j flag (such as -j 2) would get them
to pass. Rather than defaulting to a smaller -j flag value, this
commit adds a second concurrency semaphore for IntegrationJobs to
limit how many of them can run at once.

The chosen value is admittedly a bit "magic", but was chosen
because it happens to get the tests to pass on my Lenovo T480s with
a Vagrant guest.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs commented:
I marked this as WIP because I'm still testing the value to make sure it reliably works on my laptop.

abompard commented on devel/ci/bodhi-ci:
```diff
@@ -452,39 +462,47 @@ class Job(object):
+            if self.extra_concurrency_semaphore:
+                # This is a job that has stricter limitations on how many can run concurrently, so
+                # we also need to acquire the extra semaphore.
+                await self.extra_concurrency_semaphore.acquire()
```
Since Python 2.7 and 3.1 you can use [multiple context managers on the same line](https://docs.python.org/3.1/reference/compound_stmts.html#with), so you could do something like: `async with self.concurrency_semaphore, self.extra_concurrency_semaphore:`
You may however need to initialize the extra semaphore to something other than `None`, possibly using a class defining no-op `__aenter__` and `__aexit__` methods (since [contextlib.asynccontextmanager](https://docs.python.org/3.7/library/contextlib.html#contextlib.asynccontextmanager) is only available in Python 3.7).
It would save you the big try/except, but it's also a bit more code so I'm not sure it's worth it.

bowlofeggs replied:
Yeah that's not a bad idea - I'll fiddle with that a bit and see how it works. I agree it'd be nice to not use the giant ```try```/```except```.

bowlofeggs commented:
Using a value of 2 does seem to run the tests reliably on my laptop. 3 was not reliable.

bowlofeggs commented:
Hmm, I actually just saw a failure even with 2 running at once. I might need to set concurrency to 1 for it.

bowlofeggs commented:
Hmm, I've now seen it even fail with 1, possibly due to other regular tests running concurrently. I wonder if I just need to limit the overall job count to a lower number in addition to preventing too many integration tests from running concurrently…

bowlofeggs committed:
Catch an APIError when killing Greenwave.

There seems to be a frequent issue with Greenwave not working in
some way during integration testing. It seems that it may not be
starting, and one of our integration tests fails to get the IPv4
address for the Greenwave container.

When the test fails, the cleanup code runs, which tries to kill
the Greenwave container. Since the Greenwave container did not
seem to start in the first place, *this* code then fails and
produces a traceback. For some reason, this second traceback seems
to cover up the original problem, which made discovering the issue
difficult.

This commit catches the second Exception and passes. Since we
really just want to ensure that the container is stopped, it's OK
if asking it to stop raises an Exception. We can ignore it, since
that Exception means we are in the desired state anyway.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs committed:
Print Greenwave logs if it fails to start.

There was an issue[0] with the CI tests that was difficult to
debug because the Greenwave logs weren't being captured by the test
suite when it failed to start. This commit adds a startup timeout
to the Greenwave container, and prints the logs from Greenwave if
it fails to start by the timeout.

[0] https://github.com/fedora-infra/bodhi/issues/2976

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs committed:
Remove a setting from the Greenwave config.

The !FedoraAtomicCi rule is no longer supported with Greenwave 1.0.
This commit removes the references to that rule from our test
policy.

fixes #2976

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs commented:
The current failure I think is due to concurrent downloads of the database dump. I've observed that running bodhi-ci locally also concurrently downloads it multiple times instead of once. I suspect that our GroovyScipt is also downloading it more than once. I need to stop work for the day, but I think I can fix that tomorrow.

bowlofeggs commented:
Logs are [here](https://ci.centos.org/job/bodhi-pipeline/job/PR-2975/5/execution/node/584/log/?consoleFull), and you can ctrl-f for the error:

```
psycopg2.ProgrammingError: constraint "packages_stack_id_fkey" of relation "packages" does not exist
```

abompard commented:
The Jenkins log file does not show multiple downloads of the dumps, however above the error line it has:
```
22:10:26 22:09:37.581 __init__.py       DEBUG  time="2019-02-05T22:09:37Z" level=error msg="Can't add file /root/payload/devel/ci/integration/dumps/bodhi2.dump to tar: archive/tar: write too long" 
```
Since the Bodhi dump isn't huge, I think this may come from a concurrency issue too. Or because the virtual disk is full.

There may be a workaround: copy the zipped files to the postgres container and unzip them locally while doing the import. I'll add a commit to do that and you can tell me what you think.

abompard committed:
Only download the Bodhi database dump once

Signed-off-by: Aurélien Bompard <aurelien@bompard.org>

bowlofeggs commented:
@abompard Keep in mind that the log I sent you was from a single run of ```bodhi-ci```. The Groovy script actually starts several ```bodhi-ci``` processes in parallel. You can see the combined logs of all those processes [here](https://ci.centos.org/job/bodhi-pipeline/job/PR-2975/5/consoleFull). If you search there, you will see that it does get downloaded multiple times. I believe it is likely that one download was overwriting a file while it was being read by another process.

I don't think extracting the dumps in the container would necessarily fix it, because the read can still happen at the same time as a write.

The fix I'd been contemplating is to have a new ```bodhi-ci``` command that downloads the database dump and have the Groovy script run that one time before the integration tests. The other thing I had in mind was to ensure that ```bodhi-ci integration``` checks the timestamp on the download to ensure it is within a day (since we do daily backups) and skip downloading if it is (you may already have this in the code - I haven't looked closely at it yet). This second component would be nice because developers probably don't want to have to call ```download``` - it's really just a way to get the Groovy script to do it once since there are multiple processes.

Another fix is just to go back to the way it used to work - have each container download the database dump. Inefficient, but avoids these complexities. However, I think it's not too much work to do the download thing I described above, and you've already done the work to bring the process outside of the container, so I think I'd tend towards solving the single download as described above.

The other thing I've considered is to make it so the Groovy script just runs bodhi-ci all instead of running a bunch of bodhi-ci's in parallel. There are positives and negatives. On the positive, it would allow it to work in a more similar manner to how developers would use it locally. It would also allow us to have one place to manage the somewhat complex dependency chain (since this is an example of duplicate work - we have to fix it in both places). The downsides are that it would be engineering effort in the short term because we'd have to refactor the Groovy and bodhi-ci: we'd need to give bodhi-ci a machine readable output format to describe the jobs and refactor the Groovy to read that format so it can report to GitHub. Another downside is that it might be challenging to get it to "live update" the GitHub statuses as it does today (the Groovy would need to be able to live process as bodhi-ci sends updates to stdout about the progress of the various tasks.)

Anyways, I think refactoring is well outside the scope and purpose of this PR, but I thought I'd throw the idea out there to see what you think. For now, I think we just need to ensure that both bodhi-ci and the Groovy script only download the dump one time.

bowlofeggs commented on devel/ci/bodhi-ci:
```diff
@@ -1128,6 +1128,7 @@ def _build_jobs_list(
         # Download the DB dumps
         integration_deps_build_jobs.extend([
             IntegrationDumpDownloadJob(app_name="waiverdb", release="prod", buffer_output=buffer_output),
+            IntegrationDumpDownloadJob(app_name="bodhi2", release="prod", buffer_output=buffer_output),
```
+1 - I had planned to do this today ☺

abompard commented:
Right, I forgot that Groovy starts multiple bodhi-ci instances. We could also use a lockfile system (I have that almost ready), let me send you the commit and you can tell me if it's worthwhile.

bowlofeggs commented:
All of the failures were actually in reporting the results to GitHub and not in the tests. You can see that by looking at the red and green balls here:

https://ci.centos.org/job/bodhi-pipeline/job/PR-2975/7/flowGraphTable/

@abompard do you approve my commits for merging? I approve your commit, for the record.

bowlofeggs commented:
I'm going to merge this without review, because it is blocking me from reviewing other pull requests and from fixing the download issue.

abompard commented:
Post merge review approval! :-)
Sorry I missed your message.

bowlofeggs commented:
Thanks @abompard! No worries ☺
